### PR TITLE
Use new ESP-IDF I2C driver

### LINKED ISSUE #9535: ESP32-S3 I2C timeout
### CircuitPython version

```python
Adafruit CircuitPython 9.1.1 on 2024-07-22; Adafruit Feather ESP32S3 No PSRAM with ESP32S3
```

### Code/REPL

```python
import time
import busio
import board
from micropython import const
from adafruit_seesaw.seesaw import Seesaw

BUTTON_X = const(6)
BUTTON_Y = const(2)
BUTTON_A = const(5)
BUTTON_B = const(1)
BUTTON_SELECT = const(0)
BUTTON_START = const(16)
button_mask = const(
    (1 << BUTTON_X)
    | (1 << BUTTON_Y)
    | (1 << BUTTON_A)
    | (1 << BUTTON_B)
    | (1 << BUTTON_SELECT)
    | (1 << BUTTON_START)
)

i2c_bus = board.STEMMA_I2C()
seesaw = Seesaw(i2c_bus, addr=0x50)
seesaw.pin_mode_bulk(button_mask, seesaw.INPUT_PULLUP)
```

### Behavior

Traceback (most recent call last):
  File "<stdin>", line 33, in <module>
  File "adafruit_seesaw/seesaw.py", line 342, in pin_mode_bulk
  File "adafruit_seesaw/seesaw.py", line 329, in _pin_mode_bulk_x
  File "adafruit_seesaw/seesaw.py", line 520, in write
OSError: [Errno 116] ETIMEDOUT

### Description

This is the example code from adafruits gamepad qt guide. Get this everytime I try to run the code. After multiple tries, there is an error : 
Traceback (most recent call last):
  File "<stdin>", line 27, in <module>
RuntimeError: No pull up found on SDA or SCL; check your wiring

Bus this goes away if you unplug the board and power back on. On RP2040 based boards, it works out of the box (same CP version).
If I use the same hardware but with arduino code (also from gamepad qt guide), it works always.

### Additional information

_No response_

### LINKED ISSUE #9561: I2C problems on ESP32-S2 after scan
Quoting from https://github.com/adafruit/circuitpython/issues/9535#issuecomment-2297686309:

I'm having a very similar issue with I2C on QT-PY-ESP32-S2. I've come at it all kinds of ways, with one or multiple I2C devices and varying the I2C bus speed from 10_000 to 400_000.

With these i2c devices:
-     adafruit_ssd1306
-     adafruit_seesaw (rotaryio: product number 4991)

When I do a i2c.scan() it never shows *any* devices, but I can "speak" to the rotaryio. When I enable the ssd1306, I get:
```
main.py output:
sys.implementation=(name='circuitpython', version=(9, 1, 1), _machine='Adafruit QT Py ESP32S2 with ESP32S2', _mpy=262)
I2C_FREQ       : 100000
I2C_DO_SCAN    : True
I2C_USE_SSD1306: True
I2C_USE_SEESAW : True
Scanning I2C addresses: []
Position: 0
cnt_loops = 16 / sec
Traceback (most recent call last):
  File "main.py", line 69, in <module>
  File "adafruit_ssd1306.py", line 219, in show
  File "adafruit_ssd1306.py", line 287, in write_framebuf
OSError: [Errno 116] ETIMEDOUT

```

If I never do the i2c.scan(), seesaw-rotaryio and ssd1306 appear to work fine _(I've tried different instances of ssd1306 hardware)_.

In the case where I use only the seesaw-rotaryio and do i2c.scan(), the scan still shows no devices, I can communicate with the seesaw-rotario and my code speed decreases *dramatically*, _e.g. 100+ loops/second becomes ~15 loops/second._

Code below. Can anyone shed some light on this?

Thanks,
  Casa

```

# ==============================================================================
#
from micropython import const
import busio
import board
import sys
from adafruit_ticks import ticks_ms, ticks_add, ticks_less

I2C_FREQ = const(100_000)
I2C_DO_SCAN: bool = True
I2C_USE_SSD1306: bool = True
I2C_USE_SEESAW: bool = True

MS_ONE_SECOND = const(1000)

# ------------------------------------------------------------------------------
print(f"{sys.implementation=}")
print(f"I2C_FREQ       : {I2C_FREQ:6d}")
print(f"I2C_DO_SCAN    : {I2C_DO_SCAN}")
print(f"I2C_USE_SSD1306: {I2C_USE_SSD1306}")
print(f"I2C_USE_SEESAW : {I2C_USE_SEESAW}")

# ------------------------------------------------------------------------------
def print_I2C_Scan(i2c: busio.i2c) -> None:
    while not i2c.try_lock():
        pass
    print("Scanning I2C addresses:", [hex(i).upper() for i in i2c.scan()])
    i2c.unlock()

# ------------------------------------------------------------------------------
# Set up i2c0, OLED, seesaw rotaryio
#
i2c0 = busio.I2C(board.SCL, board.SDA, frequency=I2C_FREQ)

if I2C_USE_SSD1306:
    import adafruit_ssd1306
    ADDR = const(0x3C)
    WIDTH = const(128)
    HEIGHT = const(64)
    oled = adafruit_ssd1306.SSD1306_I2C(WIDTH, HEIGHT, i2c0, addr=ADDR)

if I2C_USE_SEESAW:
    from adafruit_seesaw import seesaw, rotaryio  #, digitalio
    seesaw = seesaw.Seesaw(i2c0, 0x36)
    encoder = rotaryio.IncrementalEncoder(seesaw)
    last_position = None

# ------------------------------------------------------------------------------
if I2C_DO_SCAN:
    print_I2C_Scan(i2c0)    # PROBLEM HERE

# ------------------------------------------------------------------------------
cnt_loops: int = 0
cnt_printed: int = 0
ms_target: int = ticks_add(ticks_ms(), MS_ONE_SECOND)
try:
    while True:
        cnt_loops += 1
        if not ticks_less(ticks_ms(), ms_target):

            ms_target = ticks_add(ticks_ms(), MS_ONE_SECOND)
            if cnt_printed < 10:
                cnt_printed += 1
                print(f"cnt_loops = {cnt_loops} / sec")
            cnt_loops = 0

            if I2C_USE_SSD1306:
                oled.fill(0)
                oled.show()
                # oled.line(X_MAX, Y_MAX, X_MIN, Y_MIN, 1)
                oled.vline(WIDTH // 2, 0, HEIGHT, 1)
                oled.show()

        if I2C_USE_SEESAW:
            if (position := -encoder.position) != last_position:
                last_position = position
                print("Position: {}".format(position))

except KeyboardInterrupt:
    print("Caught KeyboardInterrupt, exiting")

```

_Originally posted by @Casa-Machinalia in https://github.com/adafruit/circuitpython/issues/9535#issuecomment-2297686309_

### PR DESCRIPTION
**Pull request by dhalbert**
- Fixes #9535
- Fixes #9561 

Update to use new ESP-IDF I2C driver. @tannewt did most of this work in https://github.com/tannewt/circuitpython/commit/8a4ad54fe71b78011050f962f8a5eb7f37ffc6b5 (https://github.com/tannewt/circuitpython/tree/idf5.2_i2c). I started with that,  fixed a few things, and fixed some issues with the esp32-camera library, which recently had a PR to use the new driver.

I retested the problems reported in #9535 and #9561 amd they seem to be fixed.

I added an `I2C.probe(address)` method. There has to be a method in the class dictionary so that the native `adafruit_bus_device.I2CDevice` can do a probe for a specific address on anything that acts like `I2C`. Previously it was not actually using the common-hal `probe()` operation, but instead was "manually" doing a zero-length write. Now it uses whatever `probe()` is provided by bitbangio.I2C or busio.I2C.

In the long run the presence of `.probe()` means that uses of `.scan()` could be replaced with a tiny amount of Python code:
`[address for address in range(128) if i2c.probe(address)]`. But this incompatible change would come at a major version bump.

### CODE REVIEW

**Comment by dhalbert** [comment]
A few builds overflowed. I turned off some unused pins. I also did `CIRCUITPY_CODEOP = 0` on one or two boards. That could be done on more if necessary later.

File: ports/espressif/common-hal/espcamera/Camera.c [hunk 1]
```diff
@@ -77,13 +77,12 @@ void common_hal_espcamera_camera_construct(
 
     self->i2c = i2c;
 
-    self->camera_config.pin_pwdn = common_hal_mcu_pin_number(powerdown_pin);
-    self->camera_config.pin_reset = common_hal_mcu_pin_number(reset_pin);
-    self->camera_config.pin_xclk = common_hal_mcu_pin_number(external_clock_pin);
+    // These pins might be NULL because they were not specified.
+    self->camera_config.pin_pwdn = powerdown_pin ? common_hal_mcu_pin_number(powerdown_pin) : NO_PIN;
```
Comment (jepler) [inline ports/espressif/common-hal/espcamera/Camera.c:81]: common_hal_mcu_pin_number is implemented thus on espressif:
```
uint8_t common_hal_mcu_pin_number(const mcu_pin_obj_t *pin) {
    return pin ? pin->number : NO_PIN;
}
```
that is, the special NO_PIN value is returned for a NULL pointer input. So, I don't know why this change would be needed.

File: ports/mimxrt10xx/common-hal/busio/I2C.c [hunk 2]
```diff
@@ -177,6 +164,11 @@ void common_hal_busio_i2c_deinit(busio_i2c_obj_t *self) {
 
     self->sda = NULL;
     self->scl = NULL;
+    common_hal_busio_i2c_mark_deinit(self);
```
Comment (jepler) [inline ports/mimxrt10xx/common-hal/busio/I2C.c:167]: redundant assignment also done in mark_deinit

File: ports/mimxrt10xx/common-hal/busio/I2C.c [hunk 3]
```diff
@@ -187,9 +179,6 @@ bool common_hal_busio_i2c_probe(busio_i2c_obj_t *self, uint8_t addr) {
 }
 
 bool common_hal_busio_i2c_try_lock(busio_i2c_obj_t *self) {
-    if (common_hal_busio_i2c_deinited(self)) {
```
Comment (jepler) [inline ports/mimxrt10xx/common-hal/busio/I2C.c:190]: try_lock could be used from displayio. I think it was recently changed to relying on this returning failure if the bus was deinited. so, I don't think this can be removed.

File: ports/nordic/common-hal/busio/I2C.c [hunk 4]
```diff
@@ -176,6 +159,11 @@ void common_hal_busio_i2c_deinit(busio_i2c_obj_t *self) {
     self->scl_pin_number = NO_PIN;
```
Comment (jepler) [inline ports/nordic/common-hal/busio/I2C.c:159]: another redundant assignment after addition of mark_deinit. I'll stop noting them for now.

**Review by jepler — CHANGES_REQUESTED** [review]

**Comment by jepler** [comment]
thanks for working on this!

File: ports/broadcom/common-hal/busio/I2C.c [hunk 5]
```diff
@@ -23,28 +23,8 @@ static BSC0_Type *i2c[NUM_I2C] = {BSC0, BSC1, NULL, BSC3, BSC4, BSC5, BSC6, NULL
...
 
-void reset_i2c(void) {
-    // BSC2 is dedicated to the first HDMI output.
-    never_reset_i2c[2] = true;
-    i2c_in_use[2] = true;
-    #if BCM_VERSION == 2711
-    // BSC7 is dedicated to the second HDMI output.
-    never_reset_i2c[7] = true;
-    i2c_in_use[7] = true;
-    #endif
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/broadcom/common-hal/busio/I2C.c:37]: @tannewt I had no idea what was going on here, and just left it alone.

File: ports/espressif/common-hal/busio/I2C.c [hunk 6]
```diff
@@ -60,45 +61,47 @@ void common_hal_busio_i2c_construct(busio_i2c_obj_t *self,
...
+
+    i2c_master_bus_config_t config = {
+        .i2c_port = -1, // auto
+        .sda_io_num = sda->number,
+        .scl_io_num = scl->number,
+        .clk_source = I2C_CLK_SRC_DEFAULT,
+        .glitch_ignore_cnt = 7,
+        .flags = {
+            #if CIRCUITPY_I2C_ALLOW_INTERNAL_PULL_UP
+            .enable_internal_pullup = true, /*!< Internal GPIO pull mode for I2C sda signal*/
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/espressif/common-hal/busio/I2C.c:73]: Yes, in the old version, using the old driver, it was:
```c
      #if CIRCUITPY_I2C_ALLOW_INTERNAL_PULL_UP
        .sda_pullup_en = GPIO_PULLUP_ENABLE,  /*!< Internal GPIO pull mode for I2C sda signal*/
        .scl_pullup_en = GPIO_PULLUP_ENABLE,  /*!< Internal GPIO pull mode for I2C scl signal*/
        #else
        .sda_pullup_en = GPIO_PULLUP_DISABLE,  /*!< Internal GPIO pull mode for I2C sda signal*/
        .scl_pullup_en = GPIO_PULLUP_DISABLE,  /*!< Internal GPIO pull mode for I2C scl signal*/
        #endif
```

File: ports/espressif/common-hal/busio/I2C.c [hunk 7]
```diff
@@ -108,32 +111,27 @@ bool common_hal_busio_i2c_deinited(busio_i2c_obj_t *self) {
...
-    i2c_master_stop(cmd);
-    esp_err_t result = i2c_master_cmd_begin(self->i2c_num, cmd, timeout);
-    i2c_cmd_link_delete(cmd);
-    return result;
+    common_hal_busio_i2c_mark_deinit(self);
 }
 
 bool common_hal_busio_i2c_probe(busio_i2c_obj_t *self, uint8_t addr) {
-    esp_err_t result = i2c_zero_length_write(self, addr, pdMS_TO_TICKS(10));
+    esp_err_t result = i2c_master_probe(self->handle, addr, 1);
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/espressif/common-hal/busio/I2C.c:134]: Good catch, should be 10, because the arg is ms now.

File: ports/espressif/common-hal/espcamera/Camera.c (hunk 1 above)
Comment (dhalbert) [reply ports/espressif/common-hal/espcamera/Camera.c:81]: It's because the fields in `camera_config` are ints. `NO_PIN` is defined as `-1`, and the camera driver checks for the pin being `>= 0`. _However_, `common_hal_mcu_pin_number()` returns a `uint8_t`, so the `-1` turns into `255`, and that makes it still `>= 0`. I will put in a explicating comment.

Comment (dhalbert) [reply ports/espressif/common-hal/espcamera/Camera.c:81]: I noticed that there were GPIO errors being printed in the debug log -- that's why I fixed this.

File: lib/tinyusb
[the remark this answers is not in the document]
Comment (dhalbert) [reply lib/tinyusb:1]: oops, yes, git is bad at merging submodule updates

File: ports/mimxrt10xx/common-hal/busio/I2C.c (hunk 2 above)
Comment (dhalbert) [reply ports/mimxrt10xx/common-hal/busio/I2C.c:167]: Fixed all of these, and also added
```c
    // Ensure the object starts in its deinit state.
    common_hal_busio_i2c_mark_deinit(self);
```
at the top of each constructor. Strictly speaking, this is only necessary for `NO_PIN` deinit markers, which are -1, not 0, but for safety and consistency I did them all.

File: ports/espressif/common-hal/busio/I2C.c [hunk 8]
```diff
@@ -108,32 +111,27 @@ bool common_hal_busio_i2c_deinited(busio_i2c_obj_t *self) {
...
-static esp_err_t i2c_zero_length_write(busio_i2c_obj_t *self, uint8_t addr, TickType_t timeout) {
-    // i2c_master_write_to_device() won't do zero-length writes, so we do it by hand.
-    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
-    i2c_master_start(cmd);
-    i2c_master_write_byte(cmd, addr << 1, true);
-    i2c_master_stop(cmd);
-    esp_err_t result = i2c_master_cmd_begin(self->i2c_num, cmd, timeout);
-    i2c_cmd_link_delete(cmd);
-    return result;
+    common_hal_busio_i2c_mark_deinit(self);
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/espressif/common-hal/busio/I2C.c:130]: fixed all of these

File: ports/broadcom/common-hal/busio/I2C.c [hunk 9]
```diff
@@ -90,17 +70,21 @@ bool common_hal_busio_i2c_deinited(busio_i2c_obj_t *self) {
...
         return;
     }
-    never_reset_i2c[self->index] = false;
     i2c_in_use[self->index] = false;
 
     common_hal_reset_pin(self->sda_pin);
     common_hal_reset_pin(self->scl_pin);
     self->sda_pin = NULL;
     self->scl_pin = NULL;
+    common_hal_busio_i2c_mark_deinit(self);
```
[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/broadcom/common-hal/busio/I2C.c:87]: fixed all of these

File: ports/mimxrt10xx/common-hal/busio/I2C.c (hunk 3 above)
Comment (dhalbert) [reply ports/mimxrt10xx/common-hal/busio/I2C.c:190]: I'm not sure why this disappeared. The others were fine. Thanks for noticing this.

**Review by jepler — APPROVED** [review]
Thanks. I think this addresses all the issues I raised. I would like @tannewt 's review still on the broadcom stuff, as it looks to me like protection against resetting the HDMI outputs might have been lost.

File: ports/broadcom/common-hal/busio/I2C.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (tannewt) [reply ports/broadcom/common-hal/busio/I2C.c:37]: I think you should preserve this. These I2C peripherals are used by the GPU I think. Setting them in use prevents us from trying to use them too.

[the remark this answers is not in the document]
Comment (dhalbert) [reply ports/broadcom/common-hal/busio/I2C.c]: @tannewt I restored the `broadcom` `reset_i2c()` and its call, but could you check that this is the way you want it to be? Thanks.

**Review by tannewt — APPROVED** [review]
Looks good! Thanks!
